source/common/tcp_proxy/upstream.cc

```diff
@@ -222,7 +204,18 @@ bool HttpConnPool::valid() const { return conn_pool_ != nullptr; }
 
 void HttpConnPool::newStream(GenericConnectionPoolCallbacks* callbacks) {
   callbacks_ = callbacks;
-  upstream_ = std::make_unique<HttpUpstream>(upstream_callbacks_, hostname_);
+  switch (type_) {
```

Comment:
Please make this 
if (http1) {
http1 upstream
} else {
http2 upstream
}
and check that type_ is valid when we create, and return false to valid() if type is not valid.